cvc5.lean

```diff
@@ -1666,6 +1706,214 @@ def synthFun (solver: Solver)
+
+/-- Get the synthesis solutions of the given terms.
+
+This function should be called immediately after the solver answers *unsat* for sygus input.
+
+- `terms` The terms for which the synthesis solutions is queried.
+-/
+extern_def getSynthSolutions : Solver → (term : Array Term) → Env (Array Term)
+
+/-- Find a target term of interest using sygus enumeration, with no provided grammar.
+
+The solver will infer which grammar to use in this call, which by default will be the grammars
+specified by the function(s)-to-synthesize in the current context.
+
+SyGuS v2:
+
+```smtlib
+(find-synth :target)
+```
+
+- `fst` The identifier specifying what kind of term to find.
+
+Returns the result of the find, which is the null term if this call failed.
+
+**Warning**: this function is experimental and may change in future versions.
+-/
+private extern_def findSynthWithoutGrammar : Solver → (fst : FindSynthTarget) → Env Term
+
+/-- Find a target term of interest using sygus enumeration with a provided grammar.
+
+SyGuS v2:
+
+```smtlib
+(find-synth :target G)
+```
+
+- `fst` The identifier specifying what kind of term to find.
+- `grammar` The grammar for the term.
+
+Returns the result of the find, which is the null term if this call failed.
+
+**Warning**: this function is experimental and may change in future versions.
+-/
+private extern_def findSynthWithGrammar :
+  Solver → (fst : FindSynthTarget) → (grammar : Grammar) → Env Term
+
+/-- Find a target term of interest using sygus enumeration with an optional grammar.
+
+SyGuS v2:
+
+```smtlib
+(find-synth)
+(find-synth :target G)
+```
+
+- `fst` The identifier specifying what kind of term to find.
+- `grammar` The optional grammar for the term. If `none`, the solver will infer which grammar to use
+  in this call, which by default will be the grammars specified by the function(s)-to-synthesize in
+  the current context.
+
+Returns the result of the find, which is the null term if the call failed.
+
+**Warning**: this function is experimental and may change in future versions.
+-/
+def findSynth! (solver : Solver) (fst : FindSynthTarget)
+  (grammar : Option Grammar := none)
+: Env Term :=
+  if let some grammar := grammar
+  then solver.findSynthWithGrammar fst grammar
+  else solver.findSynthWithoutGrammar fst
+
+/-- Same as `findSynth!` but returns `none` if the call failed. -/
+def findSynth? (solver : Solver) (fst : FindSynthTarget)
+  (grammar : Option Grammar := none)
+: Env (Option Term) := do
+  let term ←
+    if let some grammar := grammar
+    then solver.findSynthWithGrammar fst grammar
+    else solver.findSynthWithoutGrammar fst
+  return if term.isNull then none else term
+
+/-- Same as `findSynth!` but throws an error if the call failed. -/
+def findSynth (solver : Solver) (fst : FindSynthTarget)
+  (grammar : Option Grammar := none)
+: Env Term := do
+  if let some term ← solver.findSynth? fst grammar
+  then return term else throw <| Error.error "call to `findSynth` failed"

```

Comment:
I don't think we should add this function. `null` results are different from `Error`.